---
title: "Intro to Transcriptomics"
author: Nicole Black, Wade Boohar, Kayla Xu
date: 07/17/22
updated: 10/18/24
---

***Deliverables***
-Upload this R Notebook to your GitHub and submit the link to your Repo on Brightspace.
-Include any graphs or figures created in this assignment in the folder with your R notebook with descriptive file names.

Since this is an optional partner activity, it is okay if your answers are the same as your partnerâ€™s as long as everyone understands it and could explain it in their own words if asked. Each person must individually push their code to Github. *At the top of your R Notebook, write the name of you and your partner(s) as a comment.*

***Complete the following coding activity and answer any following questions as comments in your R Notebook***

In SummarizedExperiment Tutorial, you learned how to manipulate the SummarizedExperiment data structure and turn it into more readable dataframes, saving them as rna_counts, rna_clinical, and rna_genes. In this semi-guided assignment, you will use these dataframes to perform differential expression analysis based on tumor status.

*Pre-Assignment*
Use knitr function to set your working directory to your analysis_data folder in 490_cluster.
```{r setup}
knitr::opts_knit$set(root.dir = normalizePath("/home1/ethenche/490_cluster/analysis_data")) 
```

If DESeq2 is not already installed, install it now
```{r}

if (!require("DESeq2", quietly = TRUE))
BiocManager::install("DESeq2")
```

Load in all necessary packages
```{r}

install.packages("DESeq2")
library(DESeq2)

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("EnhancedVolcano")

install.packages("Enhanced Volcano")
library(EnhancedVolcano)

library(TCGAbiolinks)


```



*1*
Read in the rna_clinical, rna_genes, and rna_counts dataframes which you made in the "SummarizedExperiment Guided Tutorial" R Notebook

```{r}
#rna_clinical <- as.data.frame(rna_clinical)
#rna_genes <- rna_se@rowRanges@elementMetadata
#rna_genes <- as.data.frame(rna_genes)
#rna_counts <- rna_se@assays@data$unstranded
#rna_counts <- as.data.frame(rna_counts)
rna_clinical <- read.csv("/home1/ethenche/490_cluster/analysis_data/BRCA_rna_clinical.csv")
rna_genes <- read.csv("/home1/ethenche/490_cluster/analysis_data/BRCA_rna_genes.csv")
rna_counts <- read.csv("/home1/ethenche/490_cluster/analysis_data/BRCA_rna_counts.csv")
#rna_clinical <- rna_clinical[,-25]
```


*2*
In this assignment, you will run differential expression analysis comparing patient samples by whether the sample is from a tumor or normal tissue (this is the definition column in rna_clinical). You will need to choose a variable to control for covariance of: age and/or PAM50 subtype (paper_BRCA_Subtype_PAM50). 

Manipulate those columns so that they are ready for differential expression analysis (hint: what kind of variables are they? what data type are they by default? do you need to handle unknown values?) Filter out genes with a total expression across all patients less than 1000.
```{r}
#variable of interest, I sorted patients with "Solid Tissue Normal as the normal tissue, and all other as the tumor tissue here so I can run the test"

rna_clinical$definition <- ifelse(rna_clinical$definition == "Solid Tissue Normal", 
                                   "Normal tissue", 
                                   "Tumor tissue")
rna_clinical$definition <- factor(rna_clinical$definition)
column_mask <- ifelse(colnames(rna_clinical)  %in% c("treatments", "primary_site", "disease_type"), F, T)

rna_clinical <- rna_clinical[,column_mask]
write.csv(rna_clinical, "/home1/ethenche/490_cluster/analysis_data/BRCA_rna_clinical1.csv")

#covariate
rna_clinical$age_at_diagnosis <- factor(rna_clinical$age_at_diagnosis)


#clean for NA values
definition_mask <- !is.na(rna_clinical$definition)
rna_clinical <- rna_clinical[definition_mask, ]

age_mask <- !is.na(rna_clinical$age_at_diagnosis)
rna_clinical <- rna_clinical[age_mask, ]

#edit for rna_counts and rna_genes to match the filtered rna_clinical
row.names(rna_genes) <- rna_genes$gene_id
rna_counts <- rna_se@assays@data$unstranded
colnames(rna_counts) <- rna_clinical$barcode
rownames(rna_counts) <- rna_genes$gene_id


patient_ids <- rna_clinical$barcode
rna_counts_mask <- ifelse(colnames(rna_counts) %in% patient_ids, T, F)
rna_counts_filtered <- rna_counts[, rna_counts_mask]


# pre-filtering genes








less_1000_mask <- rowSums(rna_counts_filtered) >=1000
rna_counts_filtered <- rna_counts_filtered[less_1000_mask,]
rna_genes <- rna_genes[less_1000_mask,]


```




*3*
Perform the differential expression analysis, All you need to do is fill in the appropriate # terms
```{r}
dds <- DESeqDataSetFromMatrix(countData = rna_counts_filtered,
                              colData = rna_clinical,
                              design= ~age_at_diagnosis + definition)

dds_obj <- DESeq(dds)

resultsNames(dds_obj)

results <- results(dds_obj, format = "DataFrame", 
                   contrast = c("definition", "Tumor tissue" , "Normal tissue"))

results <- data.frame(results)
```

Prepare results dataframe for EnhancedVolcano plotting. Add two columns, "-log10(padj)" and "gene_name". Fill in these columns appropriately.
```{r}
results$gene_name <- rna_genes$gene_name
results$'-log10(padj)' <- -log10(results$padj)
```

*4*
Now we will use the EnhancedVolcano package to plot our results. The code is already completed and should run without adjustment if all code up to here is correct.
```{r}
EnhancedVolcano(results,
                lab = results$gene_name,
                x = 'log2FoldChange',
                y = '-log10(padj)',
                title = 'Sample Definition: Tumor vs Normal Tissue',
                pointSize = 1.0,
                labSize = 5.0)
```

*5*
# Explain what genes from each part of the Volcano Plot mean in terms of their significance and up/down regulation. 
Since in this homework, a lot of data we are analyzing may come from the same patient and sometimes we are probably just comparing between the same patient, making the whole result not so significant. Therefore, we don't really have any significance in up/down regulation.
top-right genes: These genes have high fold changes (upregulated) and are statistically significant.
bottom-right genes: These genes are also upregulated but are not statistically significant.
top-left genes:  These genes have low expression (downregulated) but are statistically significant.
bottom-left genes: These genes are downregulated but not statistically significant.
top-middle genes: These genes have moderate fold changes (either up or down) and are statistically significant.
bottom-middle genes: These genes have moderate fold changes but lack statistical significance.

Save the picture of the volcano plot (using either ggsave() or right clicking and manually downloading the image and push this .Rmd and the image to GitHub)